[GSoC] Added the linear first order non-constant coefficient non-homogeneous solver

mijo2 opened the pull request:
#### References to other Issues or PRs
Fixes #16635
Fixes #14001 
Fixes #9204

#### Brief description of what is fixed or changed
Added linear first-order non-constant coefficient non-homogeneous solver that can handle a system with any number of equations provided the coefficient matrix is commutative with its antiderivative.

#### Other comments
The solver can handle many cases of systems of ODEs when its coefficient matrix satisfies the condition mentioned in the description. Below is an example of the case of system of ODEs handled by this solver:-
```
>>> f, g = symbols("f, g", cls=Function)
>>> x = symbols("x")
>>> eqs = [Eq(f(x).diff(x), (f(x) + g(x))*log(x) + x), Eq(g(x).diff(x), (f(x) + g(x))*log(x) + x)]
>>> dsolve(eqs)
⎡                 ⎛          ⌠                        ⎞                                       ⎛          ⌠                        ⎞                  ⎤
⎢       C₁   C₂   ⎜C₁   C₂   ⎮    -2⋅x⋅log(x) + 2⋅x   ⎟  2⋅x⋅log(x) - 2⋅x           C₁   C₂   ⎜C₁   C₂   ⎮    -2⋅x⋅log(x) + 2⋅x   ⎟  2⋅x⋅log(x) - 2⋅x⎥
⎢f(x) = ── - ── + ⎜── + ── + ⎮ x⋅ℯ                  dx⎟⋅ℯ                , g(x) = - ── + ── + ⎜── + ── + ⎮ x⋅ℯ                  dx⎟⋅ℯ                ⎥
⎣       2    2    ⎝2    2    ⌡                        ⎠                             2    2    ⎝2    2    ⌡                        ⎠                  ⎦

```


#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* solvers
  * Added the new n equations linear first-order non-constant coefficient non-homogeneous solver where the coefficient matrix of the system of ODEs is commutative with its antiderivative. 
<!-- END RELEASE NOTES -->

mijo2 committed:
solvers.ode: Added the new solver

Added the new type 4 linear solver

mijo2 committed:
solver.ode: Added documentation for the new solver

mijo2 committed:
solvers.ode: Added matching code for the new type4 solver

mijo2 committed:
solvers.ode: Added test cases for the matching function

mijo2 committed:
solvers.ode: Fixed some test cases

Some of the test cases in test_systems.py had lots of unwanted
spaces in the expression and this was removed in this commit.

mijo2 committed:
doc: Added the documentation for the Type 4 solver

mijo2 committed:
solvers.ode: Made the matching code more compact

oscarbenjamin commented on sympy/solvers/ode/ode.py:
```diff
@@ -7281,14 +7284,12 @@ def _linear_2eq_order2_type11(x, y, t, r, eq):
+        _linear_neq_order1_type4)
+
+    solver = locals()['_linear_neq_order1_{}'.format(match['type_of_equation'])]
+    sol = solver(match)
```
Using `locals` like this is a bad idea. We should not do this and should also remove the other places where this happens

mijo2 replied:
Other places, they are using `globals` instead of locals, do they have to be changed to?

mijo2 committed:
solvers.ode: Added test cases for the new solver

mijo2 commented:
@oscarbenjamin I have a test case that takes 18s to get the checksysodesol result. Is it too slow for TRAVIS?

mijo2 committed:
solvers.ode: Added regression test case for issue #16635

mijo2 committed:
solvers.ode: Added regression test cases for #9204 and #14001

oscarbenjamin commented:
I think 18 seconds is fine for now but at some point we need a review of the ODE tests. Collectively they are too slow and some of the examples are too complicated to include in the test suite.

mijo2 committed:
solvers.ode: Some minor tasks

1. Updated an incorrect test case
2. Removed travis skip for one of the test cases
3. Removed an outdated comment

mijo2 committed:
solvers.ode: Removed unwanted spaces in many of the test cases

sachin-4099 commented:
It is advisable to not make so many commits for each trivial update. It leads to long queues on Travis. Perhaps, you should stop the Travis Run on the previous commits and just make it run for the final latest commit.

mijo2 commented:
@sachin-4099 Some of those commits is useful for future reference but I get your point about building only the latest commit, I would do that. Sorry for any trouble.

mijo2 commented:
@oscarbenjamin I tried to remove the Travis skip from the test case which was taking 18s and it led to the job exceeding the maximum time limit. I think there are lots of slow test cases in `test_systems.py` which leads to this fail in the build. I would skip that test case for now.

oscarbenjamin commented:
A test can be marked as `slow` without also being skipped as too slow for travis. The job that has timed out regularly times out (#19536) and is for tests that are not marked as `slow`.

oscarbenjamin commented:
The latest changes in master that you've just merged affect matrix operations and will change the result of many of the matrix calculations used here as well as making some cases slower. This is due to #19588.

mijo2 commented:
@oscarbenjamin Yes that is correct, not only are the test cases slower, but many of them have different solutions now.

mijo2 commented:
Some of the test cases have changed solution and the old solution can be obtained by using the context manager introduced in PR #19588, for now, I will use it to get simpler forms of the solution.

mijo2 committed:
solvers.ode: Updated the test cases due to matrix dotprodsimp

Many of the test cases had to be modified to get the expected
behavior before the matrix dotprodsimp context manager was
included.

mijo2 commented:
@oscarbenjamin the checks have passed. I think if this PR looks alright then we can get this merged soon and work on improving the API for the solvers, like we can directly call the solvers by passing the coefficient matrices and required arguments.

mijo2 commented:
@oscarbenjamin Another point that I would like to mention is that can we just work with dsolve for solving the systems of ODEs with `dotprodsimp` as `True` since without it, sometimes the solutions become very complicated

oscarbenjamin commented:
> @oscarbenjamin Another point that I would like to mention is that can we just work with dsolve for solving the systems of ODEs with `dotprodsimp` as `True` since without it, sometimes the solutions become very complicated

I think we need to consider how we want the tests to work for dsolve. Some of the regression tests are very slow. Obviously it would be better if we could speed them up but at the same time we don't necessarily gain that much by testing all of these examples. We shouldn't use `dotprodsimp` in general because it can also be very slow in some cases. If it is needed to verify some of the examples from the regression tests then that is fine. Normally though tests should be designed to be simple but just complicated enough that they are not trivial so that we can see that the different possibilities seem to be handled correctly.

oscarbenjamin commented on sympy/solvers/ode/tests/test_systems.py:
```diff
@@ -594,33 +630,33 @@ def test_sysode_linear_neq_order1_type1():
-    eq14 = [Eq(f(t).diff(t), 2 * g(t) - 3 * h(t)),
-           Eq(g(t).diff(t), 4 * h(t) - 2 * f(t)),
-           Eq(h(t).diff(t), 3 * f(t) - 4 * g(t))]
+    eq14 = [Eq(f(t).diff(t), 2*g(t) - 3*h(t)),
```
Did you intend to change these spaces or is that your editor doing that?

oscarbenjamin replied:
I prefer the new spacing but I just wanted to check that it is something you are doing intentionally because it seems many unrelated tests are changed in this PR.

mijo2 replied:
> Did you intend to change these spaces or is that your editor doing that?

I changed it intentionally, as in the previous PR you mentioned how these unwanted spaces made the examples look bad.

mijo2 commented:
> We shouldn't use `dotprodsimp` in general because it can also be very slow in some cases.

Ok sure, I get it now. I am a little new to the idea of using this environment variable but I think I understand your point.

oscarbenjamin commented:
This looks good to me.

If you want to squash the commit then go ahead. You can merge it yourself.

mijo2 commented:
@oscarbenjamin I don't have write access to this repository is what GitHub is showing me, I can't merge this.

mijo2 commented:
For now, I have rebased this branch and squashed some commits.

oscarbenjamin commented:
Looks good. Thanks!
